### ISSUE TITLE
Taking Maybe value directly

### ISSUE THREAD

**Issue by megasuperlexa** [aevitas/CleanTickets#1]
https://github.com/aevitas/CleanTickets/blob/719e70b75244e7a6db7924d19e06db203647c7c3/src/CleanTickets.Domain/Maybe.cs#L20C14-L20C19

I think this is not very good as it may lead to exceptions at runtime. Better option in my opinion (and how Maybe or Option is usually done) is to have `Match()` method instead which forces the user to account for both empty value case and happy path. And the situation is resolved at compile time.

**Comment by aevitas** [comment]
Definitely see your point - what you suggest is also in line with the C# programming guidelines on that properties should always be safe to get, i.e. they are more similar to accessing a field than invoking a method. 

Vladimir Khorikov's implementation in [CSharpFunctionalExtensions](https://github.com/vkhorikov/CSharpFunctionalExtensions) takes this approach, but I found that more often than not, I don't actually have an action for the `None()` case. I do see how it would be safer to just invoke `() => {}` at runtime than throwing an `InvalidOperationException` just because some behaviour didn't net a value.

What do you typically provide for the empty value case in a `Match` call?

**Comment by megasuperlexa** [comment]
Depending on situation, e.g. the code here https://github.com/aevitas/CleanTickets/blob/719e70b75244e7a6db7924d19e06db203647c7c3/src/CleanTickets.Api/Program.cs#L40
could be:
`return result.Match( success => Results.Ok(success), () => Results.NotFound();`

But overall, this functional monadic stuff makes the whole codebase non-idiomatic, odd-looking and therefore hard to follow for ordinary C# people. 
So instead of `Maybe` I would just use the standard C# language feature of [Nullable reference types](https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references). What I typically do is turn them on as errors in the very beginning of a new project:
```
    <Nullable>enable</Nullable>
    <WarningsAsErrors>Nullable</WarningsAsErrors>
```
they're not without their issues, but are useful in the long run.
